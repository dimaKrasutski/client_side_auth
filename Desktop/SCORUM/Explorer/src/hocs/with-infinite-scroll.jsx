import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { fromEvent } from 'rxjs/observable/fromEvent';
import { filter } from 'rxjs/operators/filter';
import { map } from 'rxjs/operators/map';
import { debounceTime } from 'rxjs/operators/debounceTime';

/**
 * Using that to ensure user hits the end of document
 */
const SAFE_TRIGGER_OFFSET = 120;

const INFINITE_SCROLL_DEBOUNCE = 25;

/**
 * Having current inner height of browser window and YOffset (scrolled height from top), check if we hit end of document
 */
const isEndOfDocument = () =>
  (window.innerHeight + window.pageYOffset) >= (document.body.offsetHeight - SAFE_TRIGGER_OFFSET);

export const withInfiniteScroll = WrappedComponent =>
  class WithInfiniteScroll extends Component {
    static propTypes = {
      loadNext: PropTypes.func.isRequired,
    };

    scrollSubscription;
    scroll$ = fromEvent(window, 'scroll').pipe(
      debounceTime(INFINITE_SCROLL_DEBOUNCE),
      filter(isEndOfDocument),
      map(() => window.pageYOffset)
    );

    componentDidMount() {
      const { loadNext } = this.props;
      this.scrollSubscription = this.scroll$.subscribe(loadNext);
    }

    componentWillUnmount() {
      this.scrollSubscription.unsubscribe();
    }

    render() {
      return <WrappedComponent {...this.props} />;
    }
  };
