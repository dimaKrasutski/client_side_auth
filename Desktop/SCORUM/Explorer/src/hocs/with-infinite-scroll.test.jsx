import React from 'react';
import { withInfiniteScroll } from './with-infinite-scroll';

let wrapper;
const loadNext = jest.fn();

describe('withInfiniteScroll', () => {
  beforeEach(() => {
    const WrappedComponent = () => <div>Hey</div>;
    const WithScroll = withInfiniteScroll(WrappedComponent);
    wrapper = shallow(<WithScroll loadNext={loadNext} />);
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  /**
   * Dunno how to test window scroll yet.
   */
});
