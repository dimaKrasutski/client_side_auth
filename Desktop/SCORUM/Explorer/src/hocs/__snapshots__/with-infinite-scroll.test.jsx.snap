// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInfiniteScroll should match snapshot 1`] = `
<WrappedComponent
  loadNext={[MockFunction]}
/>
`;
